### PR TITLE
[Fix](Nereids) ReorderJoin rule cannot process MarkJoin correctly

### PR DESCRIPTION
**Pull request by zhengshiJ**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Fix two problems,
1. The logical join containing the MarkJoinSlotRefrance column will generate a plan->MarkJoinSlotreference structure when reorderJoin is executed, and the MarkJoinSlotreference column will be restored after the reorder is completed. But when filter+crossJoin exists, it will be transformed into innerJoin in the rules, causing the map to fail, and the corresponding plan cannot be found, thus losing the MarkJoinSlotreference column.
2. Originally, the MarkJoinSlotReference column was used as the NonUserVisibleOutput of logicalJoin. At the same time, when logicalApply was generated, the added logicalProject did not include the MarkJoinSlotReference column, and the invalid logicalProject was deleted based on other rules, so as to ensure that LogicalApply was under the logicalFilter and could recognize the MarkJoinSlotReference column. But there will be problems if logicalProject cannot be deleted.

Repair method
1. For logicalJoin containing MarkJoinSlotreference, the rules of reorderJoin are not executed.
2. Use MarkJoinSlotreference as the output of logicalJoin and also as the output of LogicalApply.
3. When generating LogicalApply, if MarkJoinSlotreference is included, you need to add an additional logicalProject to logicalFilter, and remove the MarkJoinSlotreference column.

eg
```
logicalFilter(subquery with disconjunct)

after SubqueryToApply

logicalProject(without markJoinSlotReference)
+-- logicalFilter(markJoinSlotReference)
    +-- logicalProject(with markJoinSlotReference)
        +-- logicalApply()
```

```
SELECT * FROM sub_query_correlated_subquery1 WHERE k1 IN (SELECT k1 FROM sub_query_correlated_subquery3) OR k1 < 10;
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Explain String                                                                                                                                                                                     |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| LogicalProject[60] ( distinct=false, projects=[k1#0, k2#1], excepts=[], canEliminate=true )                                                                                                        |
| +--LogicalProject[59] ( distinct=false, projects=[k1#0, k2#1], excepts=[], canEliminate=true )                                                                                                     |
|    +--LogicalFilter[58] ( predicates=($c$1#7#false OR (k1#0 < 10)) )                                                                                                                               |
|       +--LogicalProject[57] ( distinct=false, projects=[k1#0, k2#1, $c$1#7#false], excepts=[], canEliminate=true )                                                                                 |
|          +--LogicalApply ( correlationSlot=[], correlationFilter=Optional.empty, isMarkJoin=true, MarkJoinSlotReference=$c$1#7#false, scalarSubCorrespondingSlot=empty )                           |
|             |--LogicalOlapScan ( qualified=default_cluster:regression_test_nereids_syntax_p0.sub_query_correlated_subquery1, indexName=<index_not_selected>, selectedIndexId=63105, preAgg=ON )    |
|             +--LogicalProject[34] ( distinct=false, projects=[k1#2], excepts=[], canEliminate=true )                                                                                               |
|                +--LogicalOlapScan ( qualified=default_cluster:regression_test_nereids_syntax_p0.sub_query_correlated_subquery3, indexName=<index_not_selected>, selectedIndexId=63115, preAgg=ON ) |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

## Checklist(Required)

* [x] Does it affect the original behavior
* [x] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zhengshiJ** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.23 seconds
 stream load tsv:          439 seconds loaded 74807831229 Bytes, about 162 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          72 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230328041330_clickbench_pr_121576.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 1]
```diff
@@ -76,28 +75,31 @@
...
-            Plan plan = joinToMultiJoin(filter, planToJoinMembers);
-            Preconditions.checkState(plan instanceof MultiJoin);
-            MultiJoin multiJoin = (MultiJoin) plan;
-            ctx.statementContext.setMaxNArayInnerJoin(multiJoin.children().size());
-            Plan after = multiJoinToJoin(multiJoin, planToJoinMembers);
-            return after;
-        }).toRule(RuleType.REORDER_JOIN);
+        return logicalFilter(subTree(LogicalJoin.class, LogicalFilter.class))
+            .whenNot(filter -> filter.child() instanceof LogicalJoin
+                    && ((LogicalJoin<?, ?>) filter.child()).isMarkJoin())
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:80]: if the plan is
```
Filter()
+---Join(non mark join)
    +---Join(mark join)
```

what will happen? we always has `Project` on the top of `Filter(Mark Join)`? right?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 2]
```diff
@@ -157,6 +163,10 @@ private LogicalPlan subqueryToApply(List<SubqueryExpr> subqueryExprs, LogicalPla
         LogicalPlan tmpPlan = childPlan;
         for (int i = 0; i < subqueryExprs.size(); ++i) {
             SubqueryExpr subqueryExpr = subqueryExprs.get(i);
+            if (isNorMarkJoinExistsWithAgg(subqueryExpr, subqueryToMarkJoinSlot)) {
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:166]: isNor -> non

**Comment by morrySnow** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalApply.java [hunk 3]
```diff
@@ -131,6 +131,8 @@ public Optional<Expression> getSubCorrespondingConjunct() {
     public List<Slot> computeOutput() {
         return ImmutableList.<Slot>builder()
                 .addAll(left().getOutput())
+                .addAll(markJoinSlotReference.isPresent()
+                    ? ImmutableList.of(markJoinSlotReference.get()) : ImmutableList.of())
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalApply.java:144]: done

**Comment by zhengshiJ** [comment]
run buildall

**Comment by zhengshiJ** [comment]
run buildall

**Comment by zhengshiJ** [comment]
run p0

**Review by morrySnow — APPROVED** [review]

**Pull request merged**